homeassistant/components/template/switch.py

```diff
@@ -141,12 +142,19 @@ def template_switch_state_listener(entity, old_state, new_state):
         @callback
         def template_switch_startup(event):
             """Update template on startup."""
-            async_track_state_change(
-                self.hass, self._entities, template_switch_state_listener
-            )
+            if self._template is not None:
+                async_track_state_change(
+                    self.hass, self._entities, template_switch_state_listener
+                )
 
             self.async_schedule_update_ha_state(True)
 
+        # restore state after startup
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._state = state.state in ("true", STATE_ON)
+
         self.hass.bus.async_listen_once(
```

Comment:
We don't even need to listen for startup if there is no `self._template`. So you can do 

```python
if self._template is None:
    return

@callback
def template_switch_startup(event):
    …

self.hass.bus.async_listen_once(…)
```

Reply:
except for the state restoring should be run also (and especially) for optimistic mode. I'll move that in front of the `return` statement